### DISCUSSION TITLE
What is the technical reason for the single JVM restriction?

### DISCUSSION THREAD

**Discussion by Warhorst** [microstream-one/microstream#361] [Q&A] [answered]
Greetings,

I'm a Java developer and student, and also completely new to MicroStream. As someone who worked on multiple applications accessing third party databases, I find this technology very interesting. I'm currently researching for my master thesis (which will include MicroStream in some way) and discovered a restriction in your FAQ regarding the usage of multiple JVM processes: [doc](https://docs.microstream.one/manual/storage/faq/file-storage.html#_may_multiple_jvm_processes_access_the_same_microstream_database).

The description states that some rules must be guarded to avoid inconsistency. My questions are: 
- What exactly are these rules? Or in other words: what would possibly happen if I ignore this restriction?
- How does one process with multiple instances/threads handles this problem?

Thanks in advance!

**Comment by rdebusscher** [comment, marked as answer]
Hi,

The rule is that only a single JVM/ single PersistenceManager can write to the persistent store.  When you have multiple instances pointing to the same directory with data, they both start writing independently from each to it and will corrupt the storage.  The data on the disk will not match the Java objects in memory anymore.

When you have multiple threads modifying the object graph that is assigned the root for MicroStream, you need to apply concurrency guarantees (unless your changes are atomic like assigning a new value for a String variable).  You can make use of the ReentrantReadWriteLock class of Java and the readLock and writeLock methods, to make sure that no other threads are reading the data when you make changes.

When you have multiple instances, you must assign a 'master' and synchronize the data to all instances using for example the communication support we have added (https://docs.microstream.one/manual/communication/index.html)
The upcoming Cluster product will remove the burden of the development of this synchronization for the developer and it will support the multiple instances use-case out of the box.

Hope this clarifies your questions.

**Reply by Warhorst** [reply]
This answers my questions! Thank you for this insight.
